src/Microsoft.TestPlatform.TestHostProvider/Hosting/DotnetTestHostManager.cs

```diff
@@ -513,23 +513,115 @@ public virtual TestProcessStartInfo GetTestHostProcessStartInfo(
+                //
+                //    netcoreapp3.1 hosts only support DOTNET_ROOT and DOTNET_ROOT(x86) env variables.
+                //    net8          hosts, support also DOTNET_ROOT_<ARCH> variables, which is what we should prefer to set the location of dotnet
+                //                  in a more architecture specific way.
+                //
+                // 2) The surrounding environment might already have the environment variables set, most likely by setting DOTNET_ROOT, which is
+                //    a universal way of setting where the dotnet is, that works across all different architectures of the .NET apphost.
+                //    By setting our (hopefully more specific variable) we might overwrite what user specified, and in case of DOTNET_ROOT it is probably
+                //    preferable when we can set the DOTNET_ROOT_<ARCH> variable.
+                var testhostDllPath = Path.ChangeExtension(startInfo.FileName, ".dll");
+                // This file check is for unit tests, we expect the file to always be there. Otherwise testhost.exe would not be able to run.
+                var testhostVersionInfo = _fileHelper.Exists(testhostDllPath) ? FileVersionInfo.GetVersionInfo(testhostDllPath) : null;
+                if (testhostVersionInfo != null && testhostVersionInfo.ProductMajorPart >= 17 && testhostVersionInfo.ProductMinorPart >= 14)
+                {
+                    // This is a new testhost that builds at least against net8 we should set the architecture specific DOTNET_ROOT_<ARCH>.
+                    //
+                    // We ship just testhost.exe and testhost.x86.exe if the architecture is different we won't find the testhost*.exe and
+                    // won't reach this code, but let's write this in a generic way anyway, to avoid breaking if we add more variants of testhost*.exe.
+                    var environmentVariableName = $"DOTNET_ROOT_{_architecture.ToString().ToUpperInvariant()}";
+
+                    var existingDotnetRoot = _environmentVariableHelper.GetEnvironmentVariable(environmentVariableName);
+                    if (!StringUtilities.IsNullOrWhiteSpace(existingDotnetRoot))
+                    {
+                        // The variable is already set in the surrounding environment, don't set it, because we want to keep what user provided.
+                    }
+                    else
+                    {
+                        // Set the architecture specific variable to the environment of the process so it is picked up.
+                        startInfo.EnvironmentVariables.Add(environmentVariableName, dotnetRootPath);
+                    }
+                }
+                else
+                {
+                    // This is an old testhost that built against netcoreapp3.1, it does not understand architecture specific DOTNET_ROOT_<ARCH>, we have to set it more carefully
```

Comment:
When running old testhost (netcoreapp3.1) on 32-bit OS, I think `DOTNET_ROOT(x86)` isn't used. In this case, we should probably set `DOTNET_ROOT` (only for the case of WOW64, not generally if target architecture is x86).

Reply:
`  string rootVariableName = Environment.Is64BitProcess ? "DOTNET_ROOT" : "DOTNET_ROOT(x86)";`

This is what the code did in dotnet/sdk, so if dotnet.sdk was running as x86 it would set the x86 variable, and the old code here just took that value and replace the prefix. So we are not changing the behavior from what it was before.

so since I am not 100% sure that the DOTNET_ROOT(x86) is not used, I would keep it as it was before. That is keep the code as is in this PR.